File: helper/Sources/VideoSyncCore/ControlProtocol.swift
import Foundation

public enum ProtocolVersion {
    public static let current = 1
}

public enum CommandType: String, Codable, Sendable {
    case pair
    case configureCapture
    case startRecording
    case stopRecording
    case transferComplete
    case startWebRTCPreview
    case addWebRTCIceCandidate
    case stopWebRTCPreview
    case ping
}

public enum EventType: String, Codable, Sendable {
    case paired
    case captureConfigured
    case recordingStarted
    case recordingStopped
    case transferAcknowledged
    case webRTCPreviewAnswer
    case webRTCIceCandidateAdded
    case webRTCPreviewStopped
    case pong
    case error
}

public struct ControlCommand: Codable, Equatable, Sendable {
    public var requestID: UUID
    public var type: CommandType
    public var protocolVersion: Int
    public var token: String?
    public var pairingCode: String?
    public var sessionID: String?
    public var recordingID: String?
    public var captureProfile: CaptureProfile?
    public var webRTCOfferSDP: String?
    public var webRTCIceCandidateSDP: String?
    public var webRTCIceCandidateMid: String?
    public var webRTCIceCandidateMLineIndex: Int32?
    public var metadata: [String: String]

    public init(
        requestID: UUID = UUID(),
        type: CommandType,
        protocolVersion: Int = ProtocolVersion.current,
        token: String? = nil,
        pairingCode: String? = nil,
        sessionID: String? = nil,
        recordingID: String? = nil,
        captureProfile: CaptureProfile? = nil,
        webRTCOfferSDP: String? = nil,
        webRTCIceCandidateSDP: String? = nil,
        webRTCIceCandidateMid: String? = nil,
        webRTCIceCandidateMLineIndex: Int32? = nil,
        metadata: [String: String] = [:]
    ) {
        self.requestID = requestID
        self.type = type
        self.protocolVersion = protocolVersion
        self.token = token
        self.pairingCode = pairingCode
        self.sessionID = sessionID
        self.recordingID = recordingID
        self.captureProfile = captureProfile
        self.webRTCOfferSDP = webRTCOfferSDP
        self.webRTCIceCandidateSDP = webRTCIceCandidateSDP
        self.webRTCIceCandidateMid = webRTCIceCandidateMid
        self.webRTCIceCandidateMLineIndex = webRTCIceCandidateMLineIndex
        self.metadata = metadata
    }
}

public struct CaptureProfile: Codable, Equatable, Sendable {
    public var resolution: String
    public var fps: Int
    public var orientation: String
    public var aspectRatio: String

    public init(
        resolution: String = "4K",
        fps: Int = 30,
        orientation: String = "portrait",
        aspectRatio: String = "9:16"
    ) {
        self.resolution = resolution
        self.fps = fps
        self.orientation = orientation
        self.aspectRatio = aspectRatio
    }

    public var displayName: String {
        "\(resolution) \(fps) fps, \(orientation), \(aspectRatio)"
    }
}

public struct RecordingDescriptor: Codable, Equatable, Sendable {
    public var id: String
    public var filename: String
    public var byteCount: Int64
    public var durationSeconds: Double?
    public var checksumSHA256: String?
    public var downloadPath: String

    public init(
        id: String,
        filename: String,
        byteCount: Int64,
        durationSeconds: Double? = nil,
        checksumSHA256: String? = nil,
        downloadPath: String
    ) {
        self.id = id
        self.filename = filename
        self.byteCount = byteCount
        self.durationSeconds = durationSeconds
        self.checksumSHA256 = checksumSHA256
        self.downloadPath = downloadPath
    }
}

public struct PreviewDescriptor: Codable, Equatable, Sendable {
    public var snapshotPath: String
    public var streamPath: String
    public var binaryStreamPath: String
    public var maximumDimension: Int
    public var approximateFrameRate: Double

    public init(
        snapshotPath: String = "/preview.jpg",
        streamPath: String = "/preview.mjpg",
        binaryStreamPath: String = "/preview.bin",
        maximumDimension: Int = 640,
        approximateFrameRate: Double = 12.0
    ) {
        self.snapshotPath = snapshotPath
        self.streamPath = streamPath
        self.binaryStreamPath = binaryStreamPath
        self.maximumDimension = maximumDimension
        self.approximateFrameRate = approximateFrameRate
    }
}

public struct ControlEvent: Codable, Equatable, Sendable {
    public var requestID: UUID?
    public var type: EventType
    public var protocolVersion: Int
    public var token: String?
    public var recording: RecordingDescriptor?
    public var preview: PreviewDescriptor?
    public var captureProfile: CaptureProfile?
    public var webRTCAnswerSDP: String?
    public var message: String?

    public init(
        requestID: UUID? = nil,
        type: EventType,
        protocolVersion: Int = ProtocolVersion.current,
        token: String? = nil,
        recording: RecordingDescriptor? = nil,
        preview: PreviewDescriptor? = nil,
        captureProfile: CaptureProfile? = nil,
        webRTCAnswerSDP: String? = nil,
        message: String? = nil
    ) {
        self.requestID = requestID
        self.type = type
        self.protocolVersion = protocolVersion
        self.token = token
        self.recording = recording
        self.preview = preview
        self.captureProfile = captureProfile
        self.webRTCAnswerSDP = webRTCAnswerSDP
        self.message = message
    }
}

public enum ProtocolCodec {
    private static let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return encoder
    }()

    private static let decoder = JSONDecoder()

    public static func encodeCommand(_ command: ControlCommand) throws -> Data {
        try encoder.encode(command)
    }

    public static func decodeCommand(_ data: Data) throws -> ControlCommand {
        try decoder.decode(ControlCommand.self, from: data)
    }

    public static func encodeEvent(_ event: ControlEvent) throws -> Data {
        try encoder.encode(event)
    }

    public static func decodeEvent(_ data: Data) throws -> ControlEvent {
        try decoder.decode(ControlEvent.self, from: data)
    }
}

File: desktop/core/reaphone/control_protocol.cpp
#include "reaphone/control_protocol.h"

#include "reaphone/json.h"

#include <stdexcept>

namespace reaphone {

std::string_view commandTypeRawValue(CommandType type) {
  switch (type) {
  case CommandType::Pair:
    return "pair";
  case CommandType::ConfigureCapture:
    return "configureCapture";
  case CommandType::StartRecording:
    return "startRecording";
  case CommandType::StopRecording:
    return "stopRecording";
  case CommandType::PrepareRecording:
    return "prepareRecording";
  case CommandType::ListRecordings:
    return "listRecordings";
  case CommandType::TransferComplete:
    return "transferComplete";
  case CommandType::DeleteRecording:
    return "deleteRecording";
  case CommandType::StartWebRTCPreview:
    return "startWebRTCPreview";
  case CommandType::AddWebRTCIceCandidate:
    return "addWebRTCIceCandidate";
  case CommandType::StopWebRTCPreview:
    return "stopWebRTCPreview";
  case CommandType::Ping:
    return "ping";
  }
  return "ping";
}

namespace {

json::Value encodeCaptureProfile(const CaptureProfile &profile) {
  json::Value value = json::Value::object();
  value.set("resolution", json::Value::string(profile.resolution));
  value.set("fps", json::Value::integer(profile.fps));
  value.set("orientation", json::Value::string(profile.orientation));
  value.set("aspectRatio", json::Value::string(profile.aspectRatio));
  value.set("lens", json::Value::string(profile.lens));
  value.set("zoomFactor", json::Value::real(profile.zoomFactor));
  value.set("look", json::Value::string(profile.look));
  return value;
}

} // namespace

std::string encodeControlCommand(const ControlCommand &command) {
  json::Value root = json::Value::object();

  // Always-present, non-optional fields.
  root.set("requestID", json::Value::string(command.requestID));
  root.set("type", json::Value::string(std::string(commandTypeRawValue(command.type))));
  root.set("protocolVersion", json::Value::integer(command.protocolVersion));

  json::Value metadata = json::Value::object();
  for (const auto &[key, value] : command.metadata) {
    metadata.set(key, json::Value::string(value));
  }
  root.set("metadata", std::move(metadata));

  // Optional fields, omitted when unset (mirrors Swift encodeIfPresent).
  if (command.token) {
    root.set("token", json::Value::string(*command.token));
  }
  if (command.pairingCode) {
    root.set("pairingCode", json::Value::string(*command.pairingCode));
  }
  if (command.sessionID) {
    root.set("sessionID", json::Value::string(*command.sessionID));
  }
  if (command.recordingID) {
    root.set("recordingID", json::Value::string(*command.recordingID));
  }
  if (command.captureProfile) {
    root.set("captureProfile", encodeCaptureProfile(*command.captureProfile));
  }
  if (command.webRTCOfferSDP) {
    root.set("webRTCOfferSDP", json::Value::string(*command.webRTCOfferSDP));
  }
  if (command.webRTCIceCandidateSDP) {
    root.set("webRTCIceCandidateSDP", json::Value::string(*command.webRTCIceCandidateSDP));
  }
  if (command.webRTCIceCandidateMid) {
    root.set("webRTCIceCandidateMid", json::Value::string(*command.webRTCIceCandidateMid));
  }
  if (command.webRTCIceCandidateMLineIndex) {
    root.set("webRTCIceCandidateMLineIndex",
             json::Value::integer(*command.webRTCIceCandidateMLineIndex));
  }

  return root.dump();
}

std::optional<EventType> eventTypeFromRawValue(std::string_view rawValue) {
  if (rawValue == "paired") return EventType::Paired;
  if (rawValue == "captureConfigured") return EventType::CaptureConfigured;
  if (rawValue == "recordingStarted") return EventType::RecordingStarted;
  if (rawValue == "recordingStopped") return EventType::RecordingStopped;
  if (rawValue == "recordingPrepared") return EventType::RecordingPrepared;
  if (rawValue == "recordingsListed") return EventType::RecordingsListed;
  if (rawValue == "transferAcknowledged") return EventType::TransferAcknowledged;
  if (rawValue == "recordingDeleted") return EventType::RecordingDeleted;
  if (rawValue == "webRTCPreviewAnswer") return EventType::WebRTCPreviewAnswer;
  if (rawValue == "webRTCIceCandidateAdded") return EventType::WebRTCIceCandidateAdded;
  if (rawValue == "webRTCPreviewStopped") return EventType::WebRTCPreviewStopped;
  if (rawValue == "pong") return EventType::Pong;
  if (rawValue == "error") return EventType::Error;
  return std::nullopt;
}

namespace {

std::optional<std::string> optionalString(const json::Value &object, const std::string &key) {
  const json::Value *value = object.find(key);
  if (value && value->isString()) {
    return value->asString();
  }
  return std::nullopt;
}

std::optional<double> optionalDouble(const json::Value &object, const std::string &key) {
  const json::Value *value = object.find(key);
  if (value && value->isNumber()) {
    return value->asDouble();
  }
  return std::nullopt;
}

RecordingDescriptor decodeRecording(const json::Value &object) {
  RecordingDescriptor recording;
  if (const auto id = optionalString(object, "id")) {
    recording.id = *id;
  }
  if (const auto filename = optionalString(object, "filename")) {
    recording.filename = *filename;
  }
  if (const json::Value *byteCount = object.find("byteCount"); byteCount && byteCount->isNumber()) {
    recording.byteCount = byteCount->asInt();
  }
  recording.durationSeconds = optionalDouble(object, "durationSeconds");
  recording.checksumSHA256 = optionalString(object, "checksumSHA256");
  if (const auto downloadPath = optionalString(object, "downloadPath")) {
    recording.downloadPath = *downloadPath;
  }
  return recording;
}

PreviewDescriptor decodePreview(const json::Value &object) {
  PreviewDescriptor preview;
  if (const auto snapshotPath = optionalString(object, "snapshotPath")) {
    preview.snapshotPath = *snapshotPath;
  }
  if (const auto streamPath = optionalString(object, "streamPath")) {
    preview.streamPath = *streamPath;
  }
  if (const auto binaryStreamPath = optionalString(object, "binaryStreamPath")) {
    preview.binaryStreamPath = *binaryStreamPath;
  }
  if (const json::Value *maxDim = object.find("maximumDimension"); maxDim && maxDim->isNumber()) {
    preview.maximumDimension = static_cast<int>(maxDim->asInt());
  }
  if (const auto frameRate = optionalDouble(object, "approximateFrameRate")) {
    preview.approximateFrameRate = *frameRate;
  }
  return preview;
}

CaptureProfile decodeCaptureProfile(const json::Value &object) {
  CaptureProfile profile;
  if (const auto resolution = optionalString(object, "resolution")) {
    profile.resolution = *resolution;
  }
  if (const json::Value *fps = object.find("fps"); fps && fps->isNumber()) {
    profile.fps = static_cast<int>(fps->asInt());
  }
  if (const auto orientation = optionalString(object, "orientation")) {
    profile.orientation = *orientation;
  }
  if (const auto aspectRatio = optionalString(object, "aspectRatio")) {
    profile.aspectRatio = *aspectRatio;
  }
  if (const auto lens = optionalString(object, "lens")) {
    profile.lens = *lens;
  }
  if (const auto zoomFactor = optionalDouble(object, "zoomFactor")) {
    profile.zoomFactor = *zoomFactor;
  }
  if (const auto look = optionalString(object, "look")) {
    profile.look = *look;
  }
  return profile;
}

} // namespace

ControlEvent decodeControlEvent(std::string_view json) {
  const json::Value root = json::parse(json);
  if (!root.isObject()) {
    throw std::invalid_argument("control event must be a JSON object");
  }

  const json::Value *typeValue = root.find("type");
  if (!typeValue || !typeValue->isString()) {
    throw std::invalid_argument("control event is missing a string \"type\"");
  }

  ControlEvent event;
  event.type = eventTypeFromRawValue(typeValue->asString()).value_or(EventType::Unknown);
  event.requestID = optionalString(root, "requestID");
  if (const json::Value *version = root.find("protocolVersion"); version && version->isNumber()) {
    event.protocolVersion = static_cast<int>(version->asInt());
  }
  event.token = optionalString(root, "token");

  if (const json::Value *recording = root.find("recording"); recording && recording->isObject()) {
    event.recording = decodeRecording(*recording);
  }
  if (const json::Value *recordings = root.find("recordings"); recordings && recordings->isArray()) {
    for (const json::Value &item : recordings->items()) {
      if (item.isObject()) {
        event.recordings.push_back(decodeRecording(item));
      }
    }
  }
  if (const json::Value *preview = root.find("preview"); preview && preview->isObject()) {
    event.preview = decodePreview(*preview);
  }
  if (const json::Value *profile = root.find("captureProfile"); profile && profile->isObject()) {
    event.captureProfile = decodeCaptureProfile(*profile);
  }
  event.captureStatus = optionalString(root, "captureStatus");
  event.captureProgress = optionalDouble(root, "captureProgress");
  event.webRTCAnswerSDP = optionalString(root, "webRTCAnswerSDP");
  event.message = optionalString(root, "message");

  return event;
}

} // namespace reaphone
